sdk/identity/identity/src/credentials/authorizationCodeCredential.ts

```diff
@@ -49,19 +48,71 @@ export class AuthorizationCodeCredential implements TokenCredential {
    */
   constructor(
-    tenantId: string,
+    tenantId: string | "common",
     clientId: string,
-    clientSecret: string | undefined,
+    clientSecret: string,
     authorizationCode: string,
     redirectUri: string,
     options?: IdentityClientOptions
+  );
+    /**
+   * Creates an instance of CodeFlowCredential with the details needed
+   * to request an access token using an authentication that was obtained
+   * from Azure Active Directory.
+   *
+   * It is currently necessary for the user of this credential to initiate
+   * the authorization code flow to obtain an authorization code to be used
+   * with this credential.  A full example of this flow is provided here:
+   *
+   * https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/identity/identity/samples/authorizationCodeSample.ts
+   *
+   * @param tenantId The Azure Active Directory tenant (directory) ID or name.
+   *                 'common' may be used when dealing with multi-tenant scenarios.
+   * @param clientId The client (application) ID of an App Registration in the tenant.
+   * @param authorizationCode An authorization code that was received from following the
+                              authorization code flow.  This authorization code must not
+                              have already been used to obtain an access token.
+   * @param redirectUri The redirect URI that was used to request the authorization code.
+                        Must be the same URI that is configured for the App Registration.
+   * @param options Options for configuring the client which makes the access token request.
+   */
+  constructor(
+    tenantId: string | "common",
+    clientId: string,
+    authorizationCode: string,
+    redirectUri: string,
+    options?: IdentityClientOptions
+  ); 
+  /**
+   * @ignore
+   * @internal
+   */
+  constructor(
+    tenantId: string | "common",
+    clientId: string,
+    clientSecretOrAuthorizationCode: string,
+    authorizationCodeOrRedirectUri: string,
+    redirectUriOrOptions: string | IdentityClientOptions | undefined,
+    options?: IdentityClientOptions
   ) {
-    this.identityClient = new IdentityClient(options);
-    this.tenantId = tenantId;
     this.clientId = clientId;
-    this.clientSecret = clientSecret;
-    this.authorizationCode = authorizationCode;
-    this.redirectUri = redirectUri;
+    this.tenantId = tenantId;
+    
+    if (typeof redirectUriOrOptions === "string") {
+      // the clientId+clientSecret constructor
+      this.clientSecret = clientSecretOrAuthorizationCode;
+      this.authorizationCode = authorizationCodeOrRedirectUri;
+      this.redirectUri = redirectUriOrOptions;
+      // options okay
+    } else {
+      // clientId only
+      this.clientSecret = undefined;
+      this.authorizationCode = clientSecretOrAuthorizationCode;
+      this.redirectUri = authorizationCodeOrRedirectUri as string;      
+      options = redirectUriOrOptions as IdentityClientOptions;
```

Comment:
Do we have tests to cover both the overloads?
At the minimum, we should have tests to cover the fancy if/else above :)

Reply:
We do - there are tests that construct the object using a client secret and tests that don't for AuthorizationCodeCredential